pkg/host/features.go

```diff
@@ -16,200 +17,61 @@ import (
-	}
-	if features[FeatureSwap].Enabled {
-		result |= flatrpc.FeatureSwap
-	}
-	return result
-}
-
-var checkFeature [numFeatures]func() string
-
-func unconditionallyEnabled() string { return "" }
-
-// Check detects features supported on the host.
-// Empty string for a feature means the feature is supported,
-// otherwise the string contains the reason why the feature is not supported.
-func Check(target *prog.Target) (*Features, error) {
-	const unsupported = "support is not implemented in syzkaller"
-	res := &Features{
-		FeatureCoverage:         {Name: "code coverage", Reason: unsupported},
-		FeatureComparisons:      {Name: "comparison tracing", Reason: unsupported},
-		FeatureExtraCoverage:    {Name: "extra coverage", Reason: unsupported},
-		FeatureDelayKcovMmap:    {Name: "delay kcov mmap", Reason: unsupported},
-		FeatureSandboxSetuid:    {Name: "setuid sandbox", Reason: unsupported},
-		FeatureSandboxNamespace: {Name: "namespace sandbox", Reason: unsupported},
-		FeatureSandboxAndroid:   {Name: "Android sandbox", Reason: unsupported},
-		FeatureFault:            {Name: "fault injection", Reason: unsupported},
-		FeatureLeak:             {Name: "leak checking", Reason: unsupported},
-		FeatureNetInjection:     {Name: "net packet injection", Reason: unsupported},
-		FeatureNetDevices:       {Name: "net device setup", Reason: unsupported},
-		FeatureKCSAN:            {Name: "concurrency sanitizer", Reason: unsupported},
-		FeatureDevlinkPCI:       {Name: "devlink PCI setup", Reason: unsupported},
-		FeatureNicVF:            {Name: "NIC VF setup", Reason: unsupported},
-		FeatureUSBEmulation:     {Name: "USB emulation", Reason: unsupported},
-		FeatureVhciInjection:    {Name: "hci packet injection", Reason: unsupported},
-		FeatureWifiEmulation:    {Name: "wifi device emulation", Reason: unsupported},
-		Feature802154Emulation:  {Name: "802.15.4 emulation", Reason: unsupported},
-		FeatureSwap:             {Name: "swap file", Reason: unsupported},
+// SetupFeatures enables and does any one-time setup for the requested features on the host.
+// Note: this can be called multiple times and must be idempotent.
+func SetupFeatures(target *prog.Target, executor string, mask flatrpc.Feature, flags csource.Features) (
+	[]flatrpc.FeatureInfo, error) {
+	if noHostChecks(target) {
+		return nil, nil
 	}
-	for n, check := range featureCheckers(target) {
-		if check == nil {
+	var results []flatrpc.FeatureInfo
+	resultC := make(chan flatrpc.FeatureInfo)
+	for feat := range flatrpc.EnumNamesFeature {
+		feat := feat
+		if mask&feat == 0 {
 			continue
 		}
-		if reason := check(); reason == "" {
-			if n == FeatureCoverage && !target.ExecutorUsesShmem {
-				return nil, fmt.Errorf("enabling FeatureCoverage requires enabling ExecutorUsesShmem")
-			}
-			res[n].Enabled = true
-			res[n].Reason = "enabled"
-		} else {
-			res[n].Reason = reason
+		if feat == flatrpc.FeatureBinFmtMisc && flags != nil && !flags["binfmt_misc"].Enabled {
```

Comment:
I don't know, I am just porting the existing logic :)
Setup func in pkg/host does the same.
I guess we ought to check all csource features here that correspond to the other features, but also probably nobody care much b/c this only affects syz-execprog when you give very precise flags and want to avoid setting up a particular feature (some manual workflow).
This all is a total mess. Ideally we don't have 4 copies of the same and don't move things back and forth with no particular benefit.

Reply:
Yes, it seems that we should have checked here all the features that must be set up once. And these checks look now very similar to what we do in `ipc.FeaturesToFlags()`.

Perhaps it's already not too hard to keep the mapping of `feature id => {env flags, name}` in one place.

Reply:
Done. PTAL.

Reply:
Cool, now one less duplication. We can now merge it with `pkg/csource/options.go`'s Features in one of the next PRs.

Reply:
If we do more in this direction, we need to figure out what's the final destination, how we remove these these 5+ duplicates of the same and all of the conversion code entirely.